### PR TITLE
some fixes for macros: one esoteric, and one quite critical

### PR DESCRIPTION
**Pull request by xeno-by**
review @retronym

### CODE REVIEW

**Comment by xeno-by** [comment]
@retronym Could we please have this PR reviewed and merged in 2.10.2? It turns out even after the previous fix pointed out by Iulian there were still some always on debug thingies. Merging this PR is an easy way to improve performance of macros in 2.10.x.

File: src/reflect/scala/reflect/runtime/ReflectionUtils.scala [hunk 1]
```diff
@@ -28,9 +28,11 @@ private[scala] object ReflectionUtils {
     case ex if pf isDefinedAt unwrapThrowable(ex)   => pf(unwrapThrowable(ex))
   }
 
-  private def systemProperties: Iterator[(String, String)] = {
+  private def systemProperties: Map[String, String] = {
```
Comment (soc) [inline src/reflect/scala/reflect/runtime/ReflectionUtils.scala]: Is there a reason why you're not using `scala.util.Properties`?

I'd suggest just adding `def bootClassPath = propOrEmpty("sun.boot.class.path")` to `scala.util.Properties` and get rid of `systemProperties` and `inferBootClasspath`.

Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/ReflectionUtils.scala]: Well, I wouldn't say that `scala.util.Properties` is a good place for a non-documented API. However I agree that we should be a bit simpler here - no need to complicate our lives with property traversal and other tricks, when we can go directly to `sun.boot.class.path`.

**Comment by retronym** [comment]
LGTM.

Thanks for looking into the System Properties thing. I've actually hit the same problem once before, but forgot to file a ticket, and failed to complete a fix.

**Comment by paulp** [comment]
DNLGTM.

Really, has nobody ever yet seen sys.props?

**Comment by retronym** [comment]
Which should survive: `scala.sys.SystemProperties` or `scala.tools.reflect.WrappedProperties`? Neither is deprecated.

Whichever it is, ought to be hardened against non-string values (that's the topic of SI-7465). When dealing with isolated classloaders, it is sometimes amazing useful to have a spot for true JVM globals, and that's the only one  I know.

**Comment by paulp** [comment]
BTW I don't care if you use scala.util.Properties, my comment is in reference to seeing the manual application of JavaConverters to something which has already been pre-java-converted and in the public API since 2.8 or 2.9.

**Comment by retronym** [comment]
Isn't that in the code that is being removed?

**Comment by soc** [comment]
That's far from the only place where we have a yet-another-properties abstraction, though.

**Comment by paulp** [comment]
"Whichever it is, ought to be hardened against non-string values (that's the topic of SI-7465). When dealing with isolated classloaders, it is sometimes amazing useful to have a spot for true JVM globals, and that's the only one I know."

Do you mean that you intentionally put not-Strings in the system properties hashtable?

**Comment by retronym** [comment]
As a last resort, in tests and the like. Call me Mr Nefarious.

Most recently, I wanted to sneak a `Context` out to the next line of the REPL as captured by a macro compiled on the prior line. The other time I tried was to get something out of the Java agent in our instrumented tests.

Obviously these are total edge cases, but we still ought to harden against them.

**Comment by xeno-by** [comment]
@paulp Given that I removed the NIH wrapper for System.getProperties, does this pull request deserve your LGTM?

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 2]
```diff
@@ -341,11 +341,12 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
-    macroTraceVerbose("tparams are: ")(tparams)
-    macroTraceVerbose("vparamss are: ")(vparamss)
-    macroTraceVerbose("retTpe is: ")(retTpe)
-    macroTraceVerbose("macroImplSig is: ")((paramss, implRetTpe))
+    macroLogVerbose(s"generating macroImplSigs for: $macroDef")
+    macroLogVerbose(s"tparams are: $tparams")
+    macroLogVerbose(s"vparamss are: $vparamss")
+    macroLogVerbose(s"retTpe is: $retTpe")
+    macroLogVerbose(s"macroImplSig is: $paramss, $implRetTpe")
+    (paramss, implRetTpe)
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:351]: For future reference, generating five closures here is not very thrifty. They're inlined under -optimise, but there's no reason to depend on that. One big triple quoted string.

**Comment by paulp** [comment]
Also for the record, it's fine to harden against Properties holding non-String things, but I don't see how it's a bug. It's a java class, and java throws CCEs if you put things which aren't Strings in it. It's not exactly realistic that we can provide stronger guarantees than they can, given that its fate was sealed before we got our hands on it.

**Comment by xeno-by** [comment]
@paulp Addressed the feedback. We now account for both `java.boot.class.path` and `sun.boot.class.path`. I don't think we should try very hard, though, because that's just a debug print.

**Comment by som-snytt** [comment]
> Call me Mr Nefarious.

And now I know why they call you Mr Nefarious.

This is not news, but even as a casual outsider, I've hit the Properties duplication a couple of times. 

I mean nsc.Properties vs util.Properties vs. partest.util.Properties; at one point, two were imported in the same file; vs WrappedProperties, which I encountered in my last commit but haven't fully grokked yet.  For some reason, I needed "envOrSome".  Ack, a pun, "some" reason.

I remember a long time ago feeling miffed that PropertiesTrait was private[scala], but I've forgotten the use case.

> it is sometimes amazing useful to have a spot for true JVM globals

I haven't used it in a while, but MBeans?  Maybe they've gone out of fashion.

**Comment by xeno-by** [comment]
Can I have an LGTM please? Looks like I've addressed all the feedback, and this PR is quite critical to get into 2.10.2, as it easily boosts performance of all macro invocations.

**Comment by paulp** [comment]
You got it.

**Pull request merged**